### PR TITLE
Feature TypeOfProductCategoryFilter

### LINKED ISSUE #166: TripData / ProductCategoryFilter
At SBB a request has been made to implement a ProductCategoryFilter, along the OperatorFilter.

### PR DESCRIPTION
**Pull request by skinkie**
Fix #166

### CODE REVIEW

**Review by ue71603 — DISMISSED** [review]

File: OJP_Common.xsd [hunk 1]
```diff
@@ -72,6 +72,41 @@
 			<xs:element name="OperatorRef" type="siri:OperatorRefStructure" minOccurs="0" maxOccurs="unbounded"/>
 		</xs:sequence>
 	</xs:complexType>
+	<xs:annotation>
+		<xs:documentation>========================================= Type of Product Categories =============================================</xs:documentation>
+	</xs:annotation>
+	<xs:simpleType name="TypeOfProductCategoryType">
```
Comment (Aurige) [inline OJP_Common.xsd:78]: I guess you mean TypeOfProductCategoryIdType (to reference a TypeOfProductCategory via its Id) ?

Comment (skinkie) [inline OJP_Common.xsd:78]:
```suggestion
	<xs:simpleType name="TypeOfProductCategoryIdType">
```

File: OJP_Common.xsd [hunk 2]
```diff
@@ -72,6 +72,41 @@
...
+			<xs:documentation>Identifier of a TypeOfProductCategory</xs:documentation>
+		</xs:annotation>
+		<xs:restriction base="xs:NMTOKEN"/>
+	</xs:simpleType>
+	<xs:complexType name="TypeOfProductCategoryRefStructure">
+		<xs:annotation>
+			<xs:documentation>Reference to a TypeOfProductCategory</xs:documentation>
+		</xs:annotation>
+		<xs:simpleContent>
+			<xs:extension base="TypeOfProductCategoryType"/>
```
Comment (skinkie) [inline OJP_Common.xsd:89]:
```suggestion
			<xs:extension base="TypeOfProductCategoryIdType"/>
```

**Review by ue71603 — APPROVED** [review]

**Review by Aurige — APPROVED** [review]

**Review by sgrossberndt — COMMENTED** [review]
What is a "type of product category"? How does a TypeOfProductCategoryId differ from a ProductCategoryId? Is this really needed?

OJP_Common.xsd already defines `ProductCategoryCodeType` (also siri_feature_support-v2.0.xsd, see https://github.com/VDVde/OJP/issues/249)

In case a "type of product category" is additionally needed please add explanations to the xsd annotations, so its possible to understand the difference.

**Review by sgrossberndt — CHANGES_REQUESTED** [review]
What is a "type of product category"? How does a TypeOfProductCategoryId differ from a ProductCategoryId? Is this really needed?

OJP_Common.xsd already defines ProductCategoryCodeType (also siri_feature_support-v2.0.xsd, see https://github.com/VDVde/OJP/issues/249)

In case a "type of product category" is additionally needed please add explanations to the xsd annotations, so its possible to understand the difference.

**Comment by ue71603** [comment]
If ok also for @Aurige I will simplify this and not include new structures.

File: OJP_Common.xsd [hunk 3]
```diff
@@ -72,6 +72,41 @@
...
+		<xs:annotation>
+			<xs:documentation>Filter for in/exclusion of type of product category.</xs:documentation>
+		</xs:annotation>
+		<xs:sequence>
+			<xs:element name="Exclude" type="xs:boolean" default="true" minOccurs="0">
+				<xs:annotation>
+					<xs:documentation>Whether a type of product category in list are to include or exclude from search. Default is exclude.</xs:documentation>
+				</xs:annotation>
+			</xs:element>
+			<xs:element name="TypeOfProductCategoryRef" type="TypeOfProductCategoryRefStructure" minOccurs="0" maxOccurs="unbounded"/>
```
Comment (Aurige) [inline OJP_Common.xsd:107]: should just ref the element from line92

**Comment by Aurige** [comment]
I agree that the structure is a bit heavy and weird: you have a 	<xs:element name="TypeOfProductCategoryRef" type="TypeOfProductCategoryRefStructure"> that is never used  (line 107 should just refer it)

File: OJP_Common.xsd (hunk 3 above)
Comment (ue71603) [reply OJP_Common.xsd:107]: done

**Review by sgrossberndt — CHANGES_REQUESTED** [review]
This pull request is against master instead of changes_for_v1.1. When I tried to rebase the branch against changes_for_v1.1 I found out that branch already contains a ProductCategoryFilter from https://github.com/VDVde/OJP/pull/95. I guess this pull request and its branch are obsolete and can be removed?

**Pull request closed without merging**